### ISSUE TITLE
  logarithmic_amplitude 1 wrong result with high number frequencies

### ISSUE THREAD

**Issue by balbuze** [project-owner/PeppySpectrum#2]
I have a 230 frequency bands size spectrum.
The display is linear. I tested with simple freq files from 20Hz to 20kHz. 20,100,500,1000,2000,4000,10000,16000,20000Hz.
A 99 bands is ok with same files and display logarithmic
  ```
#------ Peppy ALSA Pipe section -------

pcm.peppy_in {
  type plug
  slave.pcm "peppy_in2"
}

pcm.peppy_in2 {
  type meter
  slave.pcm "peppy_out"
  scopes.0 peppyalsa
}

pcm_scope.peppyalsa {
  type peppyalsa
  decay_ms 400
  meter "/tmp/peppy_meter_fifo"
  meter_max 100
  meter_show 0
  spectrum "/tmp/peppy_spectrum_fifo"
  spectrum_max 100
  spectrum_size 230
  logarithmic_frequency 1
  logarithmic_amplitude 1
  smoothing_factor 2
}

pcm_scope_type.peppyalsa {
  lib /data/plugins/user_interface/peppyspectrum/alsa-lib/libpeppyalsa.so
}

#------ End of Peppy ALSA Pipe section -------
```
my spectrum
[800x480+230-230-bars-freq.zip](https://github.com/project-owner/PeppySpectrum/files/14968215/800x480%2B230-230-bars-freq.zip)

**Comment by project-owner** [comment]
Could you clarify what is the problem? Is it not showing spectrum when you have 230 bands and logarithmic amplitude?
All source files for the peppyalsa plugin which sends the data to the pipe are here:
https://github.com/project-owner/peppyalsa/tree/master/src

**Comment by balbuze** [comment]
It works, but frequencies are not displayed with a logarithmic scale. Ex 10kHz is at the middle of the screen. As a linear scale.
I designed the background with 99 bars.
In this case, it's correct and match with the image in the zip.

**Comment by project-owner** [comment]
That's probably the result of rounding issue. Though I cannot find it here:
https://github.com/project-owner/peppyalsa/blob/0544ed6f7bbc4a02049ab538636a8a5f587ca30e/src/spectrum.c#L174

**Comment by balbuze** [comment]
I can't fix or even find what is wrong. Not on my skill.
But I can confirm something is wrong...

**Comment by balbuze** [comment]
linear scale in REW
![image](https://github.com/project-owner/PeppySpectrum/assets/12850139/99f78ef1-0745-420b-b13b-fa34ecde47ef)
log scale in REW
![image](https://github.com/project-owner/PeppySpectrum/assets/12850139/56bb79a5-f265-40c1-8643-56e000ecbf7f)

**Comment by t0bit0bsen** [comment]
You reached the limits of the implementation.

The spectrum is calculated using a FFT. The result of a FFT is linear spaced. For logarithmic scale the linear results are combined to log spaced ones.

The spectrum (FFT) is calculated for 512 input values, the number of linear spaced bars you get out of the FFT is 256, which is close to 230. There is not enough data to create a suitable log-scaled output.

To increase the number of bars
`static unsigned int input_size = 512;`
has to be increased [here](https://github.com/project-owner/peppyalsa/blob/0544ed6f7bbc4a02049ab538636a8a5f587ca30e/src/spectrum.c#L47) to 1024 or 2048 and the code has to be re-compiled. By increasing the input data (must be 2 to the power of x) you also increase the processing time, but you will be able to get a log spaced spectrum with 230 bars.

**Comment by balbuze** [comment]
thanks a lot for the explanation.
I tested with different values 2048,4096,9192. But unfortunately, no change in sccale.

**Comment by t0bit0bsen** [comment]
Are you sure you compiled the code and installed the new binaries as described [here](https://github.com/project-owner/peppyalsa.doc/wiki/Installation)? After that you have to reboot your system to see the effect.

**Comment by balbuze** [comment]
Of course, I did that. With 2048: No change for my 230 bands, but worse, with factory spectrum 800x480, the max frequency is about 3kHz completely on the right side of the screen. Above, nothing is displayed (out of screen ?).

**Comment by t0bit0bsen** [comment]
Please try to change in line [133](https://github.com/project-owner/peppyalsa/blob/0544ed6f7bbc4a02049ab538636a8a5f587ca30e/src/spectrum.c#L133) `static double y_old[257];` to `static double y_old[513];` for an input size of 1024, or more general, to input_size/2 + 1;
Furthermore change in line [175](https://github.com/project-owner/peppyalsa/blob/0544ed6f7bbc4a02049ab538636a8a5f587ca30e/src/spectrum.c#L175) `double  log_gs = log10(256) / spectrum_size;` to `double  log_gs = log10(512) / spectrum_size;` or input_size/2, if you don't use 1024.

Actually I have no system to test the modifications.

Why do you need so many bars?

**Comment by balbuze** [comment]
Ok I will try on Monday.
I don't really need so many bard. But I ' m curious and wanted to see a full range spectrum. I wanted to see the accuracy in frequency too. As I wrote a plugin for Volumio, the idea is to provide some custom spectrum.
There is some other minor issues. While playing, sometimes spectrum stop and restart few seconds later.
And even with a pure one frequently ( file 10kHz) there is always a small signal shown at 20kHz. Parasite?
Thanks again!

**Comment by balbuze** [comment]
hello,
I tested with no success different values as you explain. No change in display with high number of bands.
Tested with 1024 and 2048
As soon as  static unsigned int input_size = 512; is set to 1024 or 2048, factory spectrum can't display frequencies higher than about 3kHz as I wrote above...

**Comment by balbuze** [comment]
having a deeper look, there is a change in scale with a 99 bands. In that case the max displayed frequency is a about 10kHz, completely on the right side of the screen.
